src/typechecker/FStar.TypeChecker.Cfg.fst

```diff
@@ -1063,17 +1090,104 @@ let built_in_primitive_steps : prim_step_set =
+                                rng=r }) r)
+              (fun r universes elt_t (l, lst) ->
+                 let blob = FStar.ImmutableArray.Base.of_list #term lst in
+                 Some (universes, elt_t, (l, FStar.Compiler.Dyn.mkdyn blob))),
+           NBETerm.mixed_binary_op
+             (fun (elt_t, _) -> Some elt_t)
+             (fun (l, q) ->
+               match NBETerm.arg_as_list NBETerm.e_any (l, q) with
+               | None -> None
+               | Some lst -> Some (l, lst))
+             (fun (universes, elt_t, (l, blob)) ->
+               NBETerm.mk_t <|
+               NBETerm.Lazy (Inr (blob, emb_typ EMB.(emb_typ_of e_any)),
+                             Thunk.mk (fun _ ->
+                               NBETerm.mk_t <| NBETerm.FV (S.lid_as_fv PC.immutable_array_of_list_lid S.delta_constant None,
+                                                          universes,
+                                                          [NBETerm.as_arg l]))))
+             (fun  universes elt_t (l, lst) ->
+                let blob = FStar.ImmutableArray.Base.of_list #NBETerm.t lst in
+                Some (universes, elt_t, (l, FStar.Compiler.Dyn.mkdyn blob))))
+      in
+      let arg1_as_elt_t (x:arg) : option term = Some (fst x) in
+      let arg2_as_blob (x:arg) : option FStar.Compiler.Dyn.dyn =
+          match (SS.compress (fst x)).n with
+          | Tm_lazy {blob=blob; lkind=Lazy_embedding (ET_app(head, _), _)}
+            when head=Ident.string_of_lid PC.immutable_array_t_lid -> Some blob
+          | _ -> None
+      in
+      let arg2_as_blob_nbe (x:NBETerm.arg) : option FStar.Compiler.Dyn.dyn =
+          let open FStar.TypeChecker.NBETerm in
+          match (fst x).nbe_t with
+          | Lazy (Inr (blob, ET_app(head, _)), _)
+            when head=Ident.string_of_lid PC.immutable_array_t_lid -> Some blob
+          | _ -> None
+      in
+      let length_op =
+        let embed_int (r:Range.range) (i:Z.t) : term = embed_simple EMB.e_int r i in
+        let run_op (blob:FStar.Compiler.Dyn.dyn) : option Z.t =
+            Some (BU.array_length #term (FStar.Compiler.Dyn.undyn blob))
+        in
+        ( PC.immutable_array_length_lid, 2, 1,
+          mixed_binary_op arg1_as_elt_t
+                          arg2_as_blob
+                          embed_int
+                          (fun _r _universes _ blob -> run_op blob),
+          NBETerm.mixed_binary_op
+             (fun (elt_t, _) -> Some elt_t)
+             arg2_as_blob_nbe
+             (fun (i:Z.t) -> NBETerm.embed NBETerm.e_int bogus_cbs i)
+             (fun _universes _ blob -> run_op blob) )
+      in
+      let index_op =
+          (PC.immutable_array_index_lid, 3, 1,
+           mixed_ternary_op arg1_as_elt_t
+                            arg2_as_blob
+                            arg_as_int
+                            (fun r tm -> tm)
```

Comment:
Running the primitive just returns a term (since the blob is an `IA.t term`). So, the unembedding is just the identity